File: VB/AdaptiveSlideNavigation/UserControls/Search.ascx.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web.UI.WebControls
Imports System.Xml
Imports DevExpress.Web

Namespace AdaptiveSlideNavigation.UserControls
    Partial Public Class Search
        Inherits System.Web.UI.UserControl

        Protected Sub SearchResults_Callback(ByVal sender As Object, ByVal e As CallbackEventArgsBase)
            Dim text = e.Parameter
            Dim results = DoSearch(text)
            If results.Count > 0 Then
                BindSearchResultsNavBar(results)
                SearchResultsNavBar.Visible = True
                noResultsContainer.Visible = False
            Else
                SearchResultsNavBar.Visible = False
                requestText.InnerHtml = text
                noResultsContainer.Visible = True
            End If
        End Sub

        Private Sub BindSearchResultsNavBar(ByVal data As IList(Of SearchResultItem))
            Dim group = New NavBarGroup()
            SearchResultsNavBar.Groups.Add(group)
            For Each item In data
                Dim navBarItem = New NavBarItem(text:= String.Format("<span class='title'>{0}</span> <span class='tags'>Tags: {1}</span>", item.Title, GetFromattedTags(item.Tags)), name:= "", imageUrl:= "", navigateUrl:= item.Url)
                group.Items.Add(navBarItem)
            Next item
        End Sub

        Private Function GetFromattedTags(ByVal tags As String) As String
            Return String.Concat(tags.Split(","c).Select(Function(i) String.Format("<span class='tag'>{0}</span>", i.Trim())))
        End Function

        Private Function DoSearch(ByVal text As String) As IList(Of SearchResultItem)
            Dim pagesNodes = XmlDataSource1.GetXmlDocument().GetElementsByTagName("page").OfType(Of XmlNode)()
            Return pagesNodes.Where(Function(n) n.Attributes("Title").Value.ToUpper().Contains(text.ToUpper()) OrElse n.Attributes("Tags").Value.ToUpper().Contains(text.ToUpper())).Select(Function(n) New SearchResultItem() With { _
                .Title = n.Attributes("Title").Value, _
                .Tags = n.Attributes("Tags").Value, _
                .Url = n.Attributes("Url").Value _
            }).ToList()
        End Function

        Private Class SearchResultItem
            Public Property Title() As String
            Public Property Tags() As String
            Public Property Url() As String
        End Class
    End Class
End Namespace

File: CS/AdaptiveSlideNavigation/UserControls/Search.ascx.cs
using System.Collections.Generic;
using System.Linq;
using System.Web.UI.WebControls;
using System.Xml;
using DevExpress.Web;

namespace AdaptiveSlideNavigation.UserControls {
    public partial class Search : System.Web.UI.UserControl {
        protected void SearchResults_Callback(object sender, CallbackEventArgsBase e) {
            var text = e.Parameter;
            var results = DoSearch(text);
            if(results.Count > 0) {
                BindSearchResultsNavBar(results);
                SearchResultsNavBar.Visible = true;
                noResultsContainer.Visible = false;
            } else {
                SearchResultsNavBar.Visible = false;
                requestText.InnerHtml = text;
                noResultsContainer.Visible = true;
            }
        }

        void BindSearchResultsNavBar(IList<SearchResultItem> data) {
            var group = new NavBarGroup();
            SearchResultsNavBar.Groups.Add(group);
            foreach(var item in data) {
                var navBarItem = new NavBarItem(
                    text: string.Format("<span class='title'>{0}</span> <span class='tags'>Tags: {1}</span>", item.Title, GetFromattedTags(item.Tags)),
                    name: "",
                    imageUrl: "",
                    navigateUrl: item.Url
                );
                group.Items.Add(navBarItem);
            }
        }

        string GetFromattedTags(string tags) {
            return string.Concat(tags.Split(',').Select(i => string.Format("<span class='tag'>{0}</span>", i.Trim())));
        }

        IList<SearchResultItem> DoSearch(string text) {
            var pagesNodes = XmlDataSource1.GetXmlDocument().GetElementsByTagName("page").OfType<XmlNode>();
            return pagesNodes
                .Where(n => n.Attributes["Title"].Value.ToUpper().Contains(text.ToUpper())
                    || n.Attributes["Tags"].Value.ToUpper().Contains(text.ToUpper()))
                .Select(n => new SearchResultItem() {
                    Title = n.Attributes["Title"].Value,
                    Tags = n.Attributes["Tags"].Value,
                    Url = n.Attributes["Url"].Value,
                }).ToList();
        }

        class SearchResultItem {
            public string Title { get; set; }
            public string Tags { get; set; }
            public string Url { get; set; }
        }
    }
}